cpu/rpx0xx/periph/spi.c

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright (C) 2023 Frank Engelhardt
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_rpx0xx
+ * @{
+ *
+ * @file
+ * @brief       Implementation of SPI.
+ *
+ * @author      Frank Engelhardt <frank@f9e.de>
+ *
+ * @}
+ */
+
+#include "assert.h"
+#include "bitarithm.h"
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/gpio.h"
+#include "periph/spi.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief   Allocate one lock per SPI device.
+ */
+static mutex_t locks[SPI_NUMOF];
+
+/**
+ * @brief   Save the clock prescaler values for faster bus acquisition.
+ */
+typedef struct {
+    spi_clk_t clk;
+    uint8_t cpsdvsr;
+    uint8_t scr;
+} _pl022_clk_t;
+static _pl022_clk_t pl022_clk[SPI_NUMOF];
+
+gpio_t  spi_pin_clk(spi_t spi)
+{
+    assert((unsigned)spi < SPI_NUMOF);
+    return spi_config[spi].clk_pin;
+}
+
+gpio_t  spi_pin_miso(spi_t spi)
+{
+    assert((unsigned)spi < SPI_NUMOF);
+    return spi_config[spi].miso_pin;
+}
+
+gpio_t  spi_pin_mosi(spi_t spi)
+{
+    assert((unsigned)spi < SPI_NUMOF);
+    return spi_config[spi].mosi_pin;
+}
+
+static void _poweron(spi_t spi)
+{
+    uint32_t reset_bit_mask =
+        (spi) ? RESETS_RESET_spi1_Msk : RESETS_RESET_spi0_Msk;
+
+    periph_reset(reset_bit_mask);
+    periph_reset_done(reset_bit_mask);
+}
+
+static void _poweroff(spi_t spi)
+{
+    uint32_t reset_bit_mask =
+        (spi) ? RESETS_RESET_spi1_Msk : RESETS_RESET_spi0_Msk;
+
+    periph_reset(reset_bit_mask);
+}
+
+void spi_init(spi_t spi)
+{
+    DEBUG("[rpx0xx] Call spi_init(spi=%" PRIdFAST8 ")", spi);
+
+    assert((unsigned)spi < SPI_NUMOF);
+
+    /* initialize device lock */
+    mutex_init(&locks[spi]);
+    /* trigger pin initialization */
+    spi_init_pins(spi);
+    /* clock prescaler values must be calculated */
+    pl022_clk[spi].clk = 0xff;
+}
+
+void spi_init_pins(spi_t spi)
+{
+    DEBUG("[rpx0xx] Call spi_init_pins(spi=%" PRIdFAST8 ")", spi);
+
+    assert((unsigned)spi < SPI_NUMOF);
+
+    const gpio_pad_ctrl_t mosi_pad_config = {
+        .drive_strength = DRIVE_STRENGTH_2MA,
```

Comment:
Please add a comment why this drive strength was selected

Reply:
Good point, I think I should better make the drive strength configurable via Kconfig. I have even found a hint that for higher frequencies one might want to have higher drive strength:

https://www.eevblog.com/forum/projects/mcu-io-drive-strength-vs-reliable-max-output-frequency/msg4179517/?PHPSESSID=7b16aui06epfi7e7ksbqbgks93#msg4179517

Reply:
Added comments now.